In [2]:
%cd /Volumes/FIT/lincs/

/Volumes/FIT/lincs


In [25]:
import transpose
import sframe
import re
import random
import numpy as np

Data is presented as gct file with two header lines followed by 12 sample annotation columns followed by data.  first 13 (?) rows are gene annotation rows.
first step is to extract landmark (LM) probes from imputed probes.  these should be first 978 probes.
following done in bash:
```
tudor:lincs $ tail -n +3 GSE70138_Broad_LINCS_Level4_ZSVCINF_mlr12k_n78980x22268_2015-06-30.gct | head -990 > GSE70138.annotLM.txt
```
the data is presented with genes as rows and samples as columns, the dataset is thus 990 rows X 78993 columns.  more natural to deal with transpose of matrix, so following script used to convert to 78993 x 990 matrix.  (This is slow, ~30min)

In [5]:
transpose.transpose('GSE70138.annotLM.txt')

indexing
indexed
transposing
('row', 0)
('row', 10000)
('row', 20000)
('row', 30000)
('row', 40000)
('row', 50000)
('row', 60000)
('row', 70000)


need to get rid of columns 2-13 since they interfere with import into data frame.
```
tudor:lincs $ head -1 GSE70138.annotLM.txt.transposed > GSE70138.annotLM.txt.transposed.trim
tudor:lincs $ tail -n +14 GSE70138.annotLM.txt.transposed >> GSE70138.annotLM.txt.transposed.trim
```

In [9]:
sf=sframe.SFrame.read_csv('GSE70138.annotLM.txt.transposed.trim',delimiter='\t')

PROGRESS: Finished parsing file /Volumes/FIT/lincs/GSE70138.annotLM.txt.transposed.trim
PROGRESS: Parsing completed. Parsed 100 lines in 1.36886 secs.
------------------------------------------------------
Inferred types from first line of file as 
column_type_hints=[str,str,str,str,float,str,str,str,str,int,str,str,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,float,floa

In [10]:
sf

Columns:
	id	str
	CL_Name	str
	det_plate	str
	det_well	str
	SM_Dose	float
	SM_Dose_Unit	str
	SM_LINCS_ID	str
	SM_Name	str
	SM_Center_Compound_ID	str
	SM_Time	int
	SM_Time_Unit	str
	SM_Pert_Type	str
	200814_at	float
	218597_s_at	float
	217140_s_at	float
	209253_at	float
	214404_x_at	float
	222103_at	float
	219888_at	float
	207042_at	float
	201453_x_at	float
	203627_at	float
	204087_s_at	float
	204131_s_at	float
	202388_at	float
	204418_x_at	float
	205607_s_at	float
	200059_s_at	float
	213417_at	float
	205963_s_at	float
	205067_at	float
	213702_x_at	float
	212851_at	float
	210151_s_at	float
	214435_x_at	float
	201334_s_at	float
	205811_at	float
	213721_at	float
	203512_at	float
	202627_s_at	float
	204849_at	float
	202224_at	float
	217478_s_at	float
	206254_at	float
	212591_at	float
	203336_s_at	float
	203127_s_at	float
	210337_s_at	float
	200602_at	float
	206066_s_at	float
	218245_at	float
	201589_at	float
	218113_at	float
	205581_s_at	float
	201351_s_at	float
	214274_s_at	float
	206562_s_at	float
	208687_x_at	float
	205897_at	float
	214710_s_at	float
	203135_at	float
	220091_at	float
	202908_at	float
	203358_s_at	float
	212403_at	float
	204949_at	float
	218086_at	float
	208950_s_at	float
	204531_s_at	float
	207416_s_at	float
	200882_s_at	float
	201202_at	float
	203709_at	float
	206501_x_at	float
	212955_s_at	float
	201548_s_at	float
	209040_s_at	float
	218064_s_at	float
	212189_s_at	float
	201555_at	float
	218290_at	float
	203138_at	float
	207199_at	float
	201503_at	float
	213292_s_at	float
	203154_s_at	float
	201379_s_at	float
	204514_at	float
	217872_at	float
	210038_at	float
	217789_at	float
	209572_s_at	float
	204656_at	float
	221514_at	float
	212051_at	float
	218407_x_at	float
	219581_at	float
	201011_at	float
	218619_s_at	float
	204757_s_at	float
	203007_x_at	float
	202076_at	float
	206173_x_at	float
	204756_at	float
	202635_s_at	float
	203110_at	float
	209899_s_at	float
	210788_s_at	float
	202093_s_at	float
	212614_at	float
	207515_s_at	float
	209615_s_at	float
	201614_s_at	float
	207163_s_at	float
	208728_s_at	float
	215980_s_at	float
	202318_s_at	float
	212500_at	float
	201841_s_at	float
	202735_at	float
	218584_at	float
	214259_s_at	float
	207620_s_at	float
	203338_at	float
	203171_s_at	float
	203726_s_at	float
	204297_at	float
	201170_s_at	float
	205034_at	float
	211715_s_at	float
	215000_s_at	float
	202491_s_at	float
	201811_x_at	float
	202187_s_at	float
	205120_s_at	float
	203442_x_at	float
	217956_s_at	float
	202095_s_at	float
	203291_at	float
	200843_s_at	float
	202507_s_at	float
	212723_at	float
	121_at	float
	209682_at	float
	218826_at	float
	218910_at	float
	212140_at	float
	206752_s_at	float
	217934_x_at	float
	200886_s_at	float
	209747_at	float
	209364_at	float
	200670_at	float
	201158_at	float
	219485_s_at	float
	202109_at	float
	203884_s_at	float
	202945_at	float
	202811_at	float
	204061_at	float
	206050_s_at	float
	221492_s_at	float
	202862_at	float
	204725_s_at	float
	201136_at	float
	209939_x_at	float
	201416_at	float
	218595_s_at	float
	204106_at	float
	212115_at	float
	212880_at	float
	203848_at	float
	201628_s_at	float
	204256_at	float
	212811_x_at	float
	202324_s_at	float
	221549_at	float
	204285_s_at	float
	200887_s_at	float
	206015_s_at	float
	202009_at	float
	206846_s_at	float
	219363_s_at	float
	208711_s_at	float
	219492_at	float
	201087_at	float
	201498_at	float
	217950_at	float
	202738_s_at	float
	209095_at	float
	214062_x_at	float
	207842_s_at	float
	217496_s_at	float
	205153_s_at	float
	200617_at	float
	207034_s_at	float
	209161_at	float
	204825_at	float
	200083_at	float
	201763_s_at	float
	200929_at	float
	202230_s_at	float
	205398_s_at	float
	208717_at	float
	204369_at	float
	205573_s_at	float
	204985_s_at	float
	212255_s_at	float
	218399_s_at	float
	212054_x_at	float
	201671_x_at	float
	202630_at	float
	202870_s_at	float
	218742_at	float
	200081_s_at	float
	201788_at	float
	200789_at	float
	201833_at	float
	204549_at	float
	216598_s_at	float
	203277_at	float
	203465_at	float
	40829_at	flo

In [21]:
sf.filter_by('PC3','CL_Name')

Columns:
	id	str
	CL_Name	str
	det_plate	str
	det_well	str
	SM_Dose	float
	SM_Dose_Unit	str
	SM_LINCS_ID	str
	SM_Name	str
	SM_Center_Compound_ID	str
	SM_Time	int
	SM_Time_Unit	str
	SM_Pert_Type	str
	200814_at	float
	218597_s_at	float
	217140_s_at	float
	209253_at	float
	214404_x_at	float
	222103_at	float
	219888_at	float
	207042_at	float
	201453_x_at	float
	203627_at	float
	204087_s_at	float
	204131_s_at	float
	202388_at	float
	204418_x_at	float
	205607_s_at	float
	200059_s_at	float
	213417_at	float
	205963_s_at	float
	205067_at	float
	213702_x_at	float
	212851_at	float
	210151_s_at	float
	214435_x_at	float
	201334_s_at	float
	205811_at	float
	213721_at	float
	203512_at	float
	202627_s_at	float
	204849_at	float
	202224_at	float
	217478_s_at	float
	206254_at	float
	212591_at	float
	203336_s_at	float
	203127_s_at	float
	210337_s_at	float
	200602_at	float
	206066_s_at	float
	218245_at	float
	201589_at	float
	218113_at	float
	205581_s_at	float
	201351_s_at	float
	214274_s_at	float
	206562_s_at	float
	208687_x_at	float
	205897_at	float
	214710_s_at	float
	203135_at	float
	220091_at	float
	202908_at	float
	203358_s_at	float
	212403_at	float
	204949_at	float
	218086_at	float
	208950_s_at	float
	204531_s_at	float
	207416_s_at	float
	200882_s_at	float
	201202_at	float
	203709_at	float
	206501_x_at	float
	212955_s_at	float
	201548_s_at	float
	209040_s_at	float
	218064_s_at	float
	212189_s_at	float
	201555_at	float
	218290_at	float
	203138_at	float
	207199_at	float
	201503_at	float
	213292_s_at	float
	203154_s_at	float
	201379_s_at	float
	204514_at	float
	217872_at	float
	210038_at	float
	217789_at	float
	209572_s_at	float
	204656_at	float
	221514_at	float
	212051_at	float
	218407_x_at	float
	219581_at	float
	201011_at	float
	218619_s_at	float
	204757_s_at	float
	203007_x_at	float
	202076_at	float
	206173_x_at	float
	204756_at	float
	202635_s_at	float
	203110_at	float
	209899_s_at	float
	210788_s_at	float
	202093_s_at	float
	212614_at	float
	207515_s_at	float
	209615_s_at	float
	201614_s_at	float
	207163_s_at	float
	208728_s_at	float
	215980_s_at	float
	202318_s_at	float
	212500_at	float
	201841_s_at	float
	202735_at	float
	218584_at	float
	214259_s_at	float
	207620_s_at	float
	203338_at	float
	203171_s_at	float
	203726_s_at	float
	204297_at	float
	201170_s_at	float
	205034_at	float
	211715_s_at	float
	215000_s_at	float
	202491_s_at	float
	201811_x_at	float
	202187_s_at	float
	205120_s_at	float
	203442_x_at	float
	217956_s_at	float
	202095_s_at	float
	203291_at	float
	200843_s_at	float
	202507_s_at	float
	212723_at	float
	121_at	float
	209682_at	float
	218826_at	float
	218910_at	float
	212140_at	float
	206752_s_at	float
	217934_x_at	float
	200886_s_at	float
	209747_at	float
	209364_at	float
	200670_at	float
	201158_at	float
	219485_s_at	float
	202109_at	float
	203884_s_at	float
	202945_at	float
	202811_at	float
	204061_at	float
	206050_s_at	float
	221492_s_at	float
	202862_at	float
	204725_s_at	float
	201136_at	float
	209939_x_at	float
	201416_at	float
	218595_s_at	float
	204106_at	float
	212115_at	float
	212880_at	float
	203848_at	float
	201628_s_at	float
	204256_at	float
	212811_x_at	float
	202324_s_at	float
	221549_at	float
	204285_s_at	float
	200887_s_at	float
	206015_s_at	float
	202009_at	float
	206846_s_at	float
	219363_s_at	float
	208711_s_at	float
	219492_at	float
	201087_at	float
	201498_at	float
	217950_at	float
	202738_s_at	float
	209095_at	float
	214062_x_at	float
	207842_s_at	float
	217496_s_at	float
	205153_s_at	float
	200617_at	float
	207034_s_at	float
	209161_at	float
	204825_at	float
	200083_at	float
	201763_s_at	float
	200929_at	float
	202230_s_at	float
	205398_s_at	float
	208717_at	float
	204369_at	float
	205573_s_at	float
	204985_s_at	float
	212255_s_at	float
	218399_s_at	float
	212054_x_at	float
	201671_x_at	float
	202630_at	float
	202870_s_at	float
	218742_at	float
	200081_s_at	float
	201788_at	float
	200789_at	float
	201833_at	float
	204549_at	float
	216598_s_at	float
	203277_at	float
	203465_at	float
	40829_at	flo

In [19]:
sf.save('GSE70138.annotLM.txt.transposed.trim.sf','binary')

annotate unique cell/cpd/dose/time combinations by unique 'treatid' values

In [5]:
sf=sframe.load_sframe('GSE70138.annotLM.txt.transposed.trim.sf')
sf=sf.pack_columns(['SM_LINCS_ID','SM_Dose','SM_Time','CL_Name'],new_column_name='treat')

[INFO] Start server at: ipc:///tmp/graphlab_server-532 - Server binary: /Library/Python/2.7/site-packages/sframe/unity_server - Server log: /tmp/sframe_server_1449916557.log
[INFO] GraphLab Server Version: 1.6


In [6]:
tmp=sf['treat'].unique()
tmp={tuple(k):i for i,k in enumerate(tmp)}
sf['treatid']=sf['treat'].apply(lambda x: tmp[tuple(x)])

In [7]:
sf[['treat','treatid','SM_Name']]

Columns:
	treat	list
	treatid	int
	SM_Name	str

Rows: 78980

Data:
+-------------------------------+---------+-----------+
|             treat             | treatid |  SM_Name  |
+-------------------------------+---------+-----------+
|    [DMSO, -666.0, 24, A375]   |  19565  |    DMSO   |
|    [DMSO, -666.0, 24, A375]   |  19565  |    DMSO   |
|    [DMSO, -666.0, 24, A375]   |  19565  |    DMSO   |
|    [DMSO, -666.0, 24, A375]   |  19565  |    DMSO   |
| [BRD-K76908866, 10.0, 24, ... |  20729  | CP-724714 |
| [BRD-K76908866, 3.33, 24, ... |  12583  | CP-724714 |
| [BRD-K76908866, 1.11, 24, ... |  20291  | CP-724714 |
| [BRD-K76908866, 0.37, 24, ... |  11315  | CP-724714 |
| [BRD-K76908866, 0.12, 24, ... |  19672  | CP-724714 |
| [BRD-K76908866, 0.04, 24, ... |   9877  | CP-724714 |
+-------------------------------+---------+-----------+
[78980 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

now recover original columns from 'packed' column:

In [8]:
sf=sf.unpack('treat')

In [9]:
sf.rename({'treat.0':'SM_LINCS_ID','treat.1':'SM_Dose','treat.2':'SM_Time','treat.3':'CL_Name'})

Columns:
	id	str
	det_plate	str
	det_well	str
	SM_Dose_Unit	str
	SM_Name	str
	SM_Center_Compound_ID	str
	SM_Time_Unit	str
	SM_Pert_Type	str
	200814_at	float
	218597_s_at	float
	217140_s_at	float
	209253_at	float
	214404_x_at	float
	222103_at	float
	219888_at	float
	207042_at	float
	201453_x_at	float
	203627_at	float
	204087_s_at	float
	204131_s_at	float
	202388_at	float
	204418_x_at	float
	205607_s_at	float
	200059_s_at	float
	213417_at	float
	205963_s_at	float
	205067_at	float
	213702_x_at	float
	212851_at	float
	210151_s_at	float
	214435_x_at	float
	201334_s_at	float
	205811_at	float
	213721_at	float
	203512_at	float
	202627_s_at	float
	204849_at	float
	202224_at	float
	217478_s_at	float
	206254_at	float
	212591_at	float
	203336_s_at	float
	203127_s_at	float
	210337_s_at	float
	200602_at	float
	206066_s_at	float
	218245_at	float
	201589_at	float
	218113_at	float
	205581_s_at	float
	201351_s_at	float
	214274_s_at	float
	206562_s_at	float
	208687_x_at	float
	205897_at	float
	214710_s_at	float
	203135_at	float
	220091_at	float
	202908_at	float
	203358_s_at	float
	212403_at	float
	204949_at	float
	218086_at	float
	208950_s_at	float
	204531_s_at	float
	207416_s_at	float
	200882_s_at	float
	201202_at	float
	203709_at	float
	206501_x_at	float
	212955_s_at	float
	201548_s_at	float
	209040_s_at	float
	218064_s_at	float
	212189_s_at	float
	201555_at	float
	218290_at	float
	203138_at	float
	207199_at	float
	201503_at	float
	213292_s_at	float
	203154_s_at	float
	201379_s_at	float
	204514_at	float
	217872_at	float
	210038_at	float
	217789_at	float
	209572_s_at	float
	204656_at	float
	221514_at	float
	212051_at	float
	218407_x_at	float
	219581_at	float
	201011_at	float
	218619_s_at	float
	204757_s_at	float
	203007_x_at	float
	202076_at	float
	206173_x_at	float
	204756_at	float
	202635_s_at	float
	203110_at	float
	209899_s_at	float
	210788_s_at	float
	202093_s_at	float
	212614_at	float
	207515_s_at	float
	209615_s_at	float
	201614_s_at	float
	207163_s_at	float
	208728_s_at	float
	215980_s_at	float
	202318_s_at	float
	212500_at	float
	201841_s_at	float
	202735_at	float
	218584_at	float
	214259_s_at	float
	207620_s_at	float
	203338_at	float
	203171_s_at	float
	203726_s_at	float
	204297_at	float
	201170_s_at	float
	205034_at	float
	211715_s_at	float
	215000_s_at	float
	202491_s_at	float
	201811_x_at	float
	202187_s_at	float
	205120_s_at	float
	203442_x_at	float
	217956_s_at	float
	202095_s_at	float
	203291_at	float
	200843_s_at	float
	202507_s_at	float
	212723_at	float
	121_at	float
	209682_at	float
	218826_at	float
	218910_at	float
	212140_at	float
	206752_s_at	float
	217934_x_at	float
	200886_s_at	float
	209747_at	float
	209364_at	float
	200670_at	float
	201158_at	float
	219485_s_at	float
	202109_at	float
	203884_s_at	float
	202945_at	float
	202811_at	float
	204061_at	float
	206050_s_at	float
	221492_s_at	float
	202862_at	float
	204725_s_at	float
	201136_at	float
	209939_x_at	float
	201416_at	float
	218595_s_at	float
	204106_at	float
	212115_at	float
	212880_at	float
	203848_at	float
	201628_s_at	float
	204256_at	float
	212811_x_at	float
	202324_s_at	float
	221549_at	float
	204285_s_at	float
	200887_s_at	float
	206015_s_at	float
	202009_at	float
	206846_s_at	float
	219363_s_at	float
	208711_s_at	float
	219492_at	float
	201087_at	float
	201498_at	float
	217950_at	float
	202738_s_at	float
	209095_at	float
	214062_x_at	float
	207842_s_at	float
	217496_s_at	float
	205153_s_at	float
	200617_at	float
	207034_s_at	float
	209161_at	float
	204825_at	float
	200083_at	float
	201763_s_at	float
	200929_at	float
	202230_s_at	float
	205398_s_at	float
	208717_at	float
	204369_at	float
	205573_s_at	float
	204985_s_at	float
	212255_s_at	float
	218399_s_at	float
	212054_x_at	float
	201671_x_at	float
	202630_at	float
	202870_s_at	float
	218742_at	float
	200081_s_at	float
	201788_at	float
	200789_at	float
	201833_at	float
	204549_at	float
	216598_s_at	float
	203277_at	float
	203465_at	float
	40829_at	float
	56197_at	float
	212287_at	float
	203575_at	float
	2122

In [10]:
sf[['SM_LINCS_ID','SM_Dose','SM_Time','CL_Name','treatid']]

Columns:
	SM_LINCS_ID	str
	SM_Dose	float
	SM_Time	int
	CL_Name	str
	treatid	int

Rows: 78980

Data:
+---------------+---------+---------+---------+---------+
|  SM_LINCS_ID  | SM_Dose | SM_Time | CL_Name | treatid |
+---------------+---------+---------+---------+---------+
|      DMSO     |  -666.0 |    24   |   A375  |  19565  |
|      DMSO     |  -666.0 |    24   |   A375  |  19565  |
|      DMSO     |  -666.0 |    24   |   A375  |  19565  |
|      DMSO     |  -666.0 |    24   |   A375  |  19565  |
| BRD-K76908866 |   10.0  |    24   |   A375  |  20729  |
| BRD-K76908866 |   3.33  |    24   |   A375  |  12583  |
| BRD-K76908866 |   1.11  |    24   |   A375  |  20291  |
| BRD-K76908866 |   0.37  |    24   |   A375  |  11315  |
| BRD-K76908866 |   0.12  |    24   |   A375  |  19672  |
| BRD-K76908866 |   0.04  |    24   |   A375  |   9877  |
+---------------+---------+---------+---------+---------+
[78980 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [11]:
sf.save('GSE70138.annotLM.txt.transposed.trim.sf','binary')

In [12]:
sf=sframe.load_sframe('GSE70138.annotLM.txt.transposed.trim.sf')

# exploratory analysis

In [72]:
print 'num unique cpds:{}'.format(len(sf['SM_LINCS_ID'].unique()))

num unique cpds:274


hmm.. that's a lot fewer unique cpds than I was expecting... explore experimental design further:

In [33]:
{x:len(sf[x].unique()) for x in sf.column_names() if x not in pscols}

{'CL_Name': 15,
 'SM_Center_Compound_ID': 278,
 'SM_Dose': 7,
 'SM_Dose_Unit': 2,
 'SM_LINCS_ID': 274,
 'SM_Name': 272,
 'SM_Pert_Type': 3,
 'SM_Time': 2,
 'SM_Time_Unit': 1,
 'det_plate': 216,
 'det_well': 380,
 'id': 78980,
 'treatid': 21722}

what is distribution of replicate numbers for unique treatments?

In [52]:
tmp=sf.groupby('treatid',{'count':sframe.aggregate.COUNT})
tmp.groupby('count',{'freq':sframe.aggregate.COUNT}).sort('count')

Columns:
	count	int
	freq	int

Rows: 40

Data:
+-------+-------+
| count |  freq |
+-------+-------+
|   1   |  143  |
|   2   |  1164 |
|   3   | 16620 |
|   4   |  1631 |
|   5   |  351  |
|   6   |  1413 |
|   7   |   48  |
|   8   |   79  |
|   9   |   22  |
|   10  |   2   |
+-------+-------+
[40 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

number of profiles per cell line?

In [56]:
sf.groupby(['CL_Name'],{'count':sframe.aggregate.COUNT}).sort('count',ascending=False).print_rows(num_rows=15)

+----------+-------+
| CL_Name  | count |
+----------+-------+
|   A549   |  8920 |
|   MCF7   |  8147 |
|  HCC515  |  5631 |
|   HT29   |  5621 |
|   PC3    |  5607 |
|  HEPG2   |  5604 |
|   HA1E   |  5569 |
|   A375   |  5565 |
|   BT20   |  5375 |
| MDAMB231 |  5372 |
|  HS578T  |  4422 |
|  SKBR3   |  4396 |
|  MCF10A  |  4377 |
|   HME1   |  2197 |
|  LNCAP   |  2177 |
+----------+-------+
[15 rows x 2 columns]



In [66]:
tmp=sf.groupby(['CL_Name','SM_Name','SM_Dose'],{'count':sframe.aggregate.COUNT})
tmp.sort(['CL_Name','SM_Name','SM_Dose']).print_rows(30)

+---------+---------+----------+-------+
| CL_Name | SM_Dose | SM_Name  | count |
+---------+---------+----------+-------+
|   A375  |   0.04  | (+)-JQ1  |   3   |
|   A375  |   0.12  | (+)-JQ1  |   3   |
|   A375  |   0.37  | (+)-JQ1  |   3   |
|   A375  |   1.11  | (+)-JQ1  |   3   |
|   A375  |   3.33  | (+)-JQ1  |   3   |
|   A375  |   10.0  | (+)-JQ1  |   3   |
|   A375  |   0.04  | (-)-JQ1  |   3   |
|   A375  |   0.12  | (-)-JQ1  |   3   |
|   A375  |   0.37  | (-)-JQ1  |   3   |
|   A375  |   1.11  | (-)-JQ1  |   3   |
|   A375  |   3.33  | (-)-JQ1  |   3   |
|   A375  |   10.0  | (-)-JQ1  |   3   |
|   A375  |   0.04  | A 769662 |   3   |
|   A375  |   0.12  | A 769662 |   3   |
|   A375  |   0.37  | A 769662 |   3   |
|   A375  |   1.11  | A 769662 |   3   |
|   A375  |   3.33  | A 769662 |   3   |
|   A375  |   10.0  | A 769662 |   3   |
|   A375  |   0.04  | A443654  |   3   |
|   A375  |   0.12  | A443654  |   3   |
|   A375  |   0.37  | A443654  |   3   |
|   A375  |   1.

In [67]:
sf['SM_Dose'].unique()

dtype: float
Rows: 7
[10.0, 1.11, 0.12, -666.0, 3.33, 0.37, 0.04]

In [68]:
tmp=sf.groupby(['CL_Name','SM_Name','SM_Dose'],{'count':sframe.aggregate.COUNT})
tmp.groupby('count',{'freq':sframe.aggregate.COUNT}).sort('count')

Columns:
	count	int
	freq	int

Rows: 34

Data:
+-------+-------+
| count |  freq |
+-------+-------+
|   1   |   55  |
|   2   |  346  |
|   3   | 10447 |
|   4   |   94  |
|   5   |  502  |
|   6   |  3712 |
|   7   |  937  |
|   8   |  350  |
|   9   |   15  |
|   10  |   19  |
+-------+-------+
[34 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [70]:
tmp=sf.groupby(['SM_Time'],{'count':sframe.aggregate.COUNT})
tmp

Columns:
	SM_Time	int
	count	int

Rows: 2

Data:
+---------+-------+
| SM_Time | count |
+---------+-------+
|    24   | 63107 |
|    3    | 15873 |
+---------+-------+
[2 rows x 2 columns]

In [71]:
tmp=sf.groupby(['CL_Name','SM_Name','SM_Dose','SM_Time'],{'count':sframe.aggregate.COUNT})
tmp.groupby('count',{'freq':sframe.aggregate.COUNT}).sort('count')

Columns:
	count	int
	freq	int

Rows: 40

Data:
+-------+-------+
| count |  freq |
+-------+-------+
|   1   |  142  |
|   2   |  1153 |
|   3   | 16463 |
|   4   |  1621 |
|   5   |  357  |
|   6   |  1467 |
|   7   |   59  |
|   8   |   84  |
|   9   |   29  |
|   10  |   2   |
+-------+-------+
[40 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### inference of experimental design:
looks like triplicates run at 6 doses for 272 compounds in <=15 cell lines at 24h.  additionally, ~1/4 of treatments run at 3 hrs.

# deep learning
now put together input for perturbation barcode learning.  limit to cell lines with good coverage, write out two files: z scores in one, treatid in 2nd.

#### train test split
sample 80% of treatments (skipping DMSO)

In [112]:
cpds=[x for x in sf['SM_LINCS_ID'].unique() if x != 'DMSO']
random.seed(123)
traincpds=random.sample(cpds,int(len(cpds)*.9))
testcpds=[x for x in cpds if x not in traincpds]
print 'len cpds:{}, len traincpds:{}, len testcpds:{}'.format(len(cpds),len(traincpds),len(testcpds))

len cpds:273, len traincpds:245, len testcpds:28


#### subsample
For training of replicate-identifying model, select what would be expected to be fairly distinct subset of data.  In particular, reduce 6-point dose response to 2: 10 and 1.1 uM.  Since 3hr time point underrepresented, focus on 24h.

In [108]:
#trainsamp=sf.filter_by(24,'SM_Time').filter_by([10,1.11],'SM_Dose').filter_by(traincpds,'SM_LINCS_ID')
#validsamp=sf.filter_by(24,'SM_Time').filter_by([10,1.11],'SM_Dose').filter_by(validcpds,'SM_LINCS_ID')
#testsamp=sf.filter_by(24,'SM_Time').filter_by([10,1.11],'SM_Dose').filter_by(testcpds,'SM_LINCS_ID')
#print 'train size:{}, valid:{}, test{}'.format(trainsamp.num_rows(),validsamp.num_rows(),testsamp.num_rows())
## rather than make new sframes at this point, add indicator columns to original sframe

In [113]:
sf['trainsamp']=sf['SM_LINCS_ID'].apply(lambda x: x in traincpds) * sf['SM_Dose'].apply(lambda x: x in [10,1.11]) * sf['SM_Time'].apply(lambda x: x==24)
print sf.groupby('trainsamp',{'count':sframe.aggregate.COUNT})
sf['testsamp']=sf['SM_LINCS_ID'].apply(lambda x: x in testcpds) * sf['SM_Dose'].apply(lambda x: x in [10,1.11]) * sf['SM_Time'].apply(lambda x: x==24)
print sf.groupby('testsamp',{'count':sframe.aggregate.COUNT})

+-----------+-------+
| trainsamp | count |
+-----------+-------+
|     0     | 61552 |
|     1     | 17428 |
+-----------+-------+
[2 rows x 2 columns]

+----------+-------+
| testsamp | count |
+----------+-------+
|    0     | 76750 |
|    1     |  2230 |
+----------+-------+
[2 rows x 2 columns]



In [114]:
sf.save('GSE70138.annotLM.txt.transposed.trim.sf','binary')

#### probeset columns

In [115]:
import re
pscols=[x for x in sf.column_names() if re.match('^\d+_[srx]?_?at$',x) is not None]
[x for x in sf.column_names() if x not in pscols]

['id',
 'det_plate',
 'det_well',
 'SM_Dose_Unit',
 'SM_Name',
 'SM_Center_Compound_ID',
 'SM_Time_Unit',
 'SM_Pert_Type',
 'treatid',
 'SM_LINCS_ID',
 'SM_Dose',
 'SM_Time',
 'CL_Name',
 'trainsamp',
 'testsamp']

Generate hierarchical indicator of treatments indexed as a pseud-float: &lt;cpdid>.&lt;treatid> 

In [122]:
cpdids={x:i for i,x in enumerate(sf['SM_LINCS_ID'].unique())}
sf['cpdid']=sf['SM_LINCS_ID'].apply(lambda x: cpdids[x])
sf['cpdtreatid']=sf['cpdid'].apply(lambda x: str(x))+'.'+sf['treatid'].apply(lambda x: str(x))
sf['cpdtreatid']

dtype: str
Rows: 78980
['216.19565', '216.19565', '216.19565', '216.19565', '257.20729', '257.12583', '257.20291', '257.11315', '257.19672', '257.9877', '56.17219', '56.8635', '56.4760', '56.18278', '56.4915', '56.12347', '212.7819', '212.14240', '212.15559', '212.3050', '212.14139', '216.19565', '216.19565', '216.19565', '216.19565', '178.18219', '178.16831', '178.13587', '178.18326', '178.3119', '178.2456', '170.12236', '170.5906', '170.12045', '170.1418', '170.14439', '170.2736', '97.800', '97.16444', '97.12775', '97.13559', '97.781', '97.47', '123.7642', '123.5316', '123.7915', '123.6106', '123.6838', '123.14416', '268.11858', '268.631', '268.15918', '268.9639', '268.787', '268.9855', '8.18704', '8.8586', '8.9391', '8.13002', '8.370', '8.2915', '15.19186', '15.16400', '15.30', '15.11816', '15.1944', '15.2197', '144.1997', '144.3600', '144.15088', '144.2645', '144.17318', '144.11313', '121.12531', '121.542', '121.17234', '121.1001', '121.5520', '121.678', '202.10017', '202.2752', '2

In [155]:
out=sf.filter_by(1,'trainsamp')[['cpdtreatid']+pscols]

In [156]:
out.shape

(17428, 979)

In [157]:
out[pscols].export_csv('lincstrainset.csv',header=False)

write grouping variable for deep learning model

In [158]:
#note that out['cpdtreatid'].save() generates output surrounded by quotes which is headache downstream
tmp=list(out['cpdtreatid'])
with open('lincstraingroup.csv','w') as f:
    for i in tmp:
        f.write('{}\n'.format(str(i)))

write annotations of rows

In [166]:
sf.filter_by(1,'trainsamp')[['id', 'treatid', 'cpdtreatid', 'SM_LINCS_ID', 'SM_Dose','SM_Time', 'CL_Name']].export_csv('lincstrainrowannot.csv')

write annotation of columns

In [159]:
with open('lincsdatasetcolnames.csv','w') as f:
    for i in pscols:
        f.write('{}\n'.format(str(i)))

& repeat for test data:

In [160]:
out=sf.filter_by(1,'testsamp')[['cpdtreatid']+pscols]
out[pscols].export_csv('lincstestset.csv',header=False)

In [161]:
tmp=list(out['cpdtreatid'])
with open('lincstestgroup.csv','w') as f:
    for i in tmp:
        f.write('{}\n'.format(str(i)))

sf.filter_by(1,'testsamp')[['id', 'treatid', 'cpdtreatid', 'SM_LINCS_ID', 'SM_Dose','SM_Time', 'CL_Name']].export_csv('lincstestrowannot.csv')

In [162]:
sf.save('GSE70138.annotLM.txt.transposed.trim.sf','binary')

In [169]:
sf['trainortest']=sf['trainsamp']+sf['testsamp']
sf['trainortest'].sum()

19658

In [171]:
sf.filter_by(1,'trainortest')[pscols].export_csv('lincstrainandtestset.csv',header=False)
sf.filter_by(1,'trainortest')[['id', 'treatid', 'cpdtreatid', 'SM_LINCS_ID', 'SM_Dose','SM_Time', 'CL_Name']].export_csv('lincstrainandtestrowannot.csv')
sf[pscols].export_csv('lincsallsampset.csv',header=False)
sf[['id', 'treatid', 'cpdtreatid', 'SM_LINCS_ID', 'SM_Dose','SM_Time', 'CL_Name']].export_csv('lincsallsamprowannot.csv')

#### model training
Model trained using the siamesenet module and models specified using siamese1.yaml or derivatives:
```
$ ./runsiamese.py siamese1.yaml
```
Hyperparameters around previous optimal solution explored.  Optimal model achieves F1 score of 0.92

#### sample encoding
Encoding script used to encode entire dataset as perturbation barcodes with parameters learned form current dataset:
```
$ python siamesemodelencode.py siamese1_best.pkl lincsallsampset.csv 
```
